src/gui/createNanotubeSpeciesDialog.cpp

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2025 Team Dissolve and contributors
+
+#include "gui/createNanotubeSpeciesDialog.h"
+#include "classes/empiricalFormula.h"
+#include "gui/helpers/comboPopulator.h"
+
+CreateNanotubeSpeciesDialog::CreateNanotubeSpeciesDialog(QWidget *parent, Dissolve &dissolve)
+    : QDialog(parent), selectElementDialog_(this), dissolve_(dissolve)
+{
+    ui_.setupUi(this);
+
+    Locker refreshLock(widgetsUpdating_);
+
+    ui_.ElementAButton->setText(QString::fromStdString(std::string(Elements::symbol(zA_))));
+    ui_.ElementBButton->setText(QString::fromStdString(std::string(Elements::symbol(zB_))));
+
+    // Set display configuration
+    ui_.StructureViewer->setSpecies(&species_);
+
+    refreshLock.unlock();
+
+    regenerate();
+}
+
+/*
+ * Data
+ */
+
+// Plot an AB atom layer
+void CreateNanotubeSpeciesDialog::plotLayer(double z, double tubeRadius, double radialStep, double radialOffset,
+                                            Elements::Element zA, Elements::Element zB)
+{
+    for (auto radial = 0; radial < ui_.RadialRingSizeSpin->value(); ++radial)
+    {
+        auto angle = radial * radialStep + radialOffset;
+
+        if (zA != Elements::Unknown)
+            species_.addAtom(zA, {tubeRadius * cos(angle), tubeRadius * sin(angle), z});
+
+        angle += radialStep * 0.5;
+
+        if (zB != Elements::Unknown)
+            species_.addAtom(
+                zB, {tubeRadius * cos(angle), tubeRadius * sin(angle), z + ui_.BondLengthSpin->value() * cos(60.0 / DEGRAD)});
+    }
```

Comment:
```suggestion
    auto angles = std::views::iota(0, ui_.RadialRingSizeSpin->value()) |
                  std::views::transform([=](const auto i) { return i * radialStep + radialOffset; });
    if (zA != Elements::Unknown)
        std::for_each(angles.cbegin(), angles.cend(),
                      [&](auto angle) { species_.addAtom(zA, {tubeRadius * cos(angle), tubeRadius * sin(angle), z}); });

    if (zB != Elements::Unknown)
        std::for_each(angles.cbegin(), angles.cend(),
                      [&](auto angle)
                      {
                          species_.addAtom(zB, {tubeRadius * cos(angle + radialStep * 0.5),
                                                tubeRadius * sin(angle + radialStep * 0.5),
                                                z + ui_.BondLengthSpin->value() * cos(60.0 / DEGRAD)});
                      });
```

Just a suggestion (and I actually think this looks *worse*), but I noticed some new features in C++ `ranges` such as pipe operators could help generate the angles.

Reply:
Nice use of `iota` and `transform` but I would agree that this rather complicates the working of the function so will stick with the original dumb loop.